verifiers/utils/dsse.go

```diff
@@ -43,3 +90,127 @@ func DecodeSignature(s string) ([]byte, error) {
 
 	return nil, fmt.Errorf("%w: %v", serrors.ErrorInvalidEncoding, errs)
 }
+
+type SignatureEncoding int
+
+const (
+	// The DER signature is encoded using ASN.1
+	// (https://tools.ietf.org/html/rfc5480#appendix-A):
+	// ECDSA-Sig-Value :: = SEQUENCE { r INTEGER, s INTEGER }. In particular, the
+	// encoding is:
+	// 0x30 || totalLength || 0x02 || r's length || r || 0x02 || s's length || s.
+	SignatureEncodingDER SignatureEncoding = iota
+	// The IEEE_P1363 signature's format is r || s, where r and s are zero-padded
+	// and have the same size in bytes as the order of the curve. For example, for
+	// NIST P-256 curve, r and s are zero-padded to 32 bytes.
+	SignatureEncodingIEEE_P1363
+)
+
+type publicKey struct {
+	keyID    string
+	pubKey   *crypto.PublicKey
+	encoding SignatureEncoding // Default is SignatureEncodingDER.
+}
+
+func (p *publicKey) Verify(ctx context.Context, data, sig []byte) error {
+	digest := sha256.Sum256(data)
+	if p.pubKey == nil {
+		return fmt.Errorf("%w: key is empty", serrors.ErrorInternal)
+	}
+	switch v := (*p.pubKey).(type) {
+	default:
+		return fmt.Errorf("unknown key type: %T", v)
+	case *ecdsa.PublicKey:
+		switch p.encoding {
+		case SignatureEncodingDER:
+			if !ecdsa.VerifyASN1(v, digest[:], sig) {
+				return fmt.Errorf("%w: cannot verify signature",
+					serrors.ErrorInvalidSignature)
+			}
+		case SignatureEncodingIEEE_P1363:
+			r := new(big.Int)
+			r.SetBytes(sig[:32])
+			s := new(big.Int)
+			s.SetBytes(sig[32:])
+			if !ecdsa.Verify(v, digest[:], r, s) {
+				return fmt.Errorf("%w: cannot verify signature",
+					serrors.ErrorInvalidSignature)
+			}
+		default:
+			return fmt.Errorf("unsupported encoding: %v", p.encoding)
+		}
+	}
+	return nil
+}
+
+// KeyID implements dsse.Verifier.KeyID.
+func (p *publicKey) KeyID() (string, error) {
+	return p.keyID, nil
+}
+
+// Public implements dsse.Verifier.Public.
+func (p *publicKey) Public() crypto.PublicKey {
+	return p.pubKey
+}
+
+type DsseVerifier struct {
+	pubKey   publicKey
+	verifier *dsselib.EnvelopeVerifier
+}
+
+type KeyFormat int
+
+const (
+	KeyFormatDER KeyFormat = iota
+	KeyFormatPEM
+)
+
+func DsseVerifierNew(content []byte, format KeyFormat, keyID string) (*DsseVerifier, error) {
+	if format == KeyFormatPEM {
+		block, _ := pem.Decode(content)
+		if block == nil {
+			return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidPEM, content)
```

Comment:
nit: not sure we should include the entire public key contents in the error message.
```suggestion
			return nil, fmt.Errorf("%w: unable to decode PEM format", serrors.ErrorInvalidPEM)
```